### PR TITLE
fix(colab): Add linter for malformed lists in generated notebooks

### LINKED ISSUE #3946: [BUG] - Incorrect list rendering in Google Colab tutorials
### Add Link

1. https://docs.pytorch.org/tutorials/advanced/usb_semisup_learn.html
  2. https://docs.pytorch.org/tutorials/beginner/basics/autogradqs_tutorial.html
  3. https://docs.pytorch.org/tutorials/beginner/basics/data_tutorial.html
  4. https://docs.pytorch.org/tutorials/beginner/blitz/neural_networks_tutorial.html
  5. https://docs.pytorch.org/tutorials/beginner/nn_tutorial.html
  6. https://docs.pytorch.org/tutorials/intermediate/optimizer_step_in_backward_tutorial.html
  7. https://docs.pytorch.org/tutorials/recipes/recipes/profiler_recipe.html
  8. https://docs.pytorch.org/tutorials/recipes/recipes/timer_quick_start.html
  9. https://docs.pytorch.org/tutorials/recipes/torch_compiler_set_stance_tutorial.html
  10. https://docs.pytorch.org/tutorials/unstable/gpu_direct_storage.html

### Description

The list-rendering problem previously reported in #3939 also occurs in several other tutorials.

On the PyTorch Tutorials HTML pages, the affected lists render correctly. However, after opening the generated notebooks through **Run in Google Colab**, the same content is not rendered as a proper list.

  ### Steps to reproduce

  1. Open one of the tutorial pages listed below.
  2. Select **Run in Google Colab**.
  3. Navigate to the affected section.
  4. Compare the list in Google Colab with the corresponding list on the tutorial HTML page.

  ### Affected tutorials

  #### 1. USB Semi-Supervised Learning

  Tutorial: https://docs.pytorch.org/tutorials/advanced/usb_semisup_learn.html

  Affected content: the list of functions imported from `semilearn` in the **Use USB to Train FreeMatch/SoftMatch on CIFAR-10 with only 40 labels** section.

Image

  #### 2. Automatic Differentiation with `torch.autograd`

  Tutorial: https://docs.pytorch.org/tutorials/beginner/basics/autogradqs_tutorial.html

  Affected content:

  - the list in the note explaining gradient availability; and
  - the list of reasons for disabling gradient tracking.

Image

  #### 3. Datasets & DataLoaders

  Tutorial: https://docs.pytorch.org/tutorials/beginner/basics/data_tutorial.html

  Affected content: the FashionMNIST parameter list in the **Loading a Dataset** section.

Image

  #### 4. Neural Networks

  Tutorial: https://docs.pytorch.org/tutorials/beginner/blitz/neural_networks_tutorial.html

  Affected content: the lists under **Recap**, **At this point, we covered**, and **Still Left**.

Image

  #### 5. What is `torch.nn` really?

  Tutorial: https://docs.pytorch.org/tutorials/beginner/nn_tutorial.html

  Affected content: the list of assumptions in the **Wrapping DataLoader** section.

Image

  #### 6. Optimizer Step in Backward

  Tutorial: https://docs.pytorch.org/tutorials/intermediate/optimizer_step_in_backward_tutorial.html

  Affected content: the numbered list following **Several major observations**.

Image

  #### 7. PyTorch Profiler

  Tutorial: https://docs.pytorch.org/tutorials/recipes/recipes/profiler_recipe.html

  Affected content: the nested activity types under the `activities` profiler parameter in the **Using profiler to analyze execution time** section.

Image

  #### 8. Timer Quick Start

  Tutorial: https://docs.pytorch.org/tutorials/recipes/recipes/timer_quick_start.html

  Affected content: the **Contents** list near the beginning of the tutorial.

Image

  #### 9. Changing the Compilation Stance

  Tutorial: https://docs.pytorch.org/tutorials/recipes/torch_compiler_set_stance_tutorial.html

  Affected content: the list following **Other stances include**.

Image

  #### 10. GPU Direct Storage

  Tutorial: https://docs.pytorch.org/tutorials/unstable/gpu_direct_storage.html

  Affected content: the list following **The steps involved in the process are as follows**.

Image

  ### Expected result

  Lists in the generated Google Colab notebooks should preserve the same structure as the corresponding lists on the PyTorch Tutorials HTML pages:

  - each item should appear on a separate line;
  - bullet and numbered lists should retain their markers;
  - nested items should retain their hierarchy; and
  - surrounding prose should remain separate from the list.

### Describe your environment

Google Colab

### PR DESCRIPTION
**Pull request by ribhuji**
Fixes #3946 

## Description
Some list structures in Sphinx-Gallery tutorial sources render correctly in the HTML documentation but become malformed after conversion to Markdown in the generated Google Colab notebooks.

This PR adds a repository-local `TUTORIAL_MARKUP` Lintrunner check that detects the two source patterns responsible for the problems documented in #3946:

- indented lists that begin without a separating blank narrative line; and
- list continuation lines that are not indented beneath their list item.

The checker only examines Sphinx-Gallery narrative content from module docstrings and gallery comment blocks. It accounts for valid nested lists, reStructuredText directives, and list tables to avoid reporting those as violations.

### Validation strategy

This draft is being built in two commits so the Lintrunner behavior can be verified directly in GitHub Actions:

1. Add and register the linter.
     The expected CI result is 14 diagnostics across exactly the 10 tutorials documented in #3946, with no additional tutorial files reported.
2. Fix the detected violations.
     The subsequent CI run should pass with no `TUTORIAL_MARKUP` diagnostics.

### Local validation

```bash
python3 -m unittest discover -s tools/linter/tests -p 'test_*.py'
lintrunner --all-files --take TUTORIAL_MARKUP
```
The unit test suite currently passes all eight tests. The complete linter scan reports 14 diagnostics across the 10 affected tutorial files and reports no other tutorial files.

## Checklist

- [x] The issue that is being fixed is referred in the description (see above "Fixes #ISSUE_NUMBER")
- [x] Only one issue is addressed in this pull request
- [x] Labels from the issue that this PR is fixing are added to this pull request
- [x] No unnecessary issues are included into this pull request.

### CODE REVIEW

**Comment by ribhuji** [comment]
@albanD, could you please approve the workflow run for this draft PR?

The current commit intentionally contains only the new Lintrunner check. The job is expected to fail with 14 diagnostics across the 10 tutorials documented in #3946, which will validate that the check detects the existing issues.

Once that run is recorded, I’ll push a second commit fixing those violations. This will preserve both the detection run and the final passing run in the PR history. Thank you!

**Comment by ribhuji** [comment]
Hey @albanD, bumping this up!

**Comment by albanD** [comment]
Sorry, I don't really have a lot of time to review things here. What is this one about?

**Comment by ribhuji** [comment]
14 Colabs created from the tutorials have indentation problems, adding a Lintrunner check for that.

**Comment by albanD** [comment]
Why is it not fixing these notebooks?

**Comment by ribhuji** [comment]
I don't just want to fix these issues. But I also want to add a check that stops these issues from occurring in any new tutorials as well.

If we run the workflow now, the workflow job fails with 14 diagnostics across the 10 tutorials documented in https://github.com/pytorch/tutorials/issues/3946, which will validate that the Lintrunner detects the existing issues, and there is value in having the Lintrunner check being added to this repo.

After that I'll push a second commit to this PR fixing these issues, the lintrunner will pass then. This will preserve both the detection run and the final passing run in the PR history.

**Comment by ribhuji** [comment]
### 1. USB Semi-Supervised Learning

Tutorial: https://docs.pytorch.org/tutorials/advanced/usb_semisup_learn.html

Fixed content: the list of functions imported from semilearn in the Use USB to Train FreeMatch/SoftMatch on CIFAR-10 with only 40 labels section.

Colab - https://colab.research.google.com/drive/1pLl7Z1-vo22Z_d9wKxoBsPM6zv9pnpZz?usp=sharing

image

### 2. Automatic Differentiation with torch.autograd

Tutorial: https://docs.pytorch.org/tutorials/beginner/basics/autogradqs_tutorial.html

Fixed content:

* the list in the note explaining gradient availability; and
* the list of reasons for disabling gradient tracking.

Colab - https://colab.research.google.com/drive/106DTdB7xXC7ICHb4kv4cvY9GvPFooH3n?usp=sharing

image

### 3. Datasets & DataLoaders

Tutorial: https://docs.pytorch.org/tutorials/beginner/basics/data_tutorial.html

Fixed content: the FashionMNIST parameter list in the Loading a Dataset section.

Colab - https://colab.research.google.com/drive/1Ryu6JY2FgQJqdqwzn5UPSuqz5RoPi2mY?usp=sharing

image

### 4. Neural Networks
Tutorial: https://docs.pytorch.org/tutorials/beginner/blitz/neural_networks_tutorial.html

Fixed content: the lists under Recap, At this point, we covered, and Still Left.

Colab - https://colab.research.google.com/drive/1kOz58dE3EyV2BzFO_0bJZeh9-ZQ3S2VC?usp=sharing

image

### 5. What is torch.nn really?
Tutorial: https://docs.pytorch.org/tutorials/beginner/nn_tutorial.html

Fixed content: the list of assumptions in the Wrapping DataLoader section.

Colab - https://colab.research.google.com/drive/15YAO3JGZueMIJRPydokTHfpv8gNa0nDX?usp=sharing

image

### 6. Optimizer Step in Backward
Tutorial: https://docs.pytorch.org/tutorials/intermediate/optimizer_step_in_backward_tutorial.html

Fixed content: the numbered list following Several major observations.

Colab - https://colab.research.google.com/drive/1ZYB3KAJUllybwpv6XoYQgvi_DI1j-zDQ?usp=sharing

image

### 7. PyTorch Profiler
Tutorial: https://docs.pytorch.org/tutorials/recipes/recipes/profiler_recipe.html

Fixed content: the nested activity types under the activities profiler parameter in the Using profiler to analyze execution time section.

Colab - https://colab.research.google.com/drive/153edt0JVmGtAkl-C4_Mtbq8xFRlRwd4h?usp=sharing

image


### 8. Timer Quick Start
Tutorial: https://docs.pytorch.org/tutorials/recipes/recipes/timer_quick_start.html

Fixed content: the Contents list near the beginning of the tutorial.

Colab - https://colab.research.google.com/drive/1kjvktsPU5fUun1WURCT5nrOfCaAQ7Zdm?usp=sharing

image


### 9. Changing the Compilation Stance
Tutorial: https://docs.pytorch.org/tutorials/recipes/torch_compiler_set_stance_tutorial.html

Fixed content: the list following Other stances include.

Colab - https://colab.research.google.com/drive/1Qz2BziA-WoU5A511VkPTSfOnFKI-28zE?usp=sharing

image


### 10. GPU Direct Storage
Tutorial: https://docs.pytorch.org/tutorials/unstable/gpu_direct_storage.html

Fixed content: the list following The steps involved in the process are as follows.

Colab - https://colab.research.google.com/drive/1XZDmHG-fy7LhldeIFwIODijuT9fKRdrt?usp=sharing

image


------

Built all 10 with - 
```
GALLERY_PATTERN='(?:^|/)(?:usb_semisup_learn|autogradqs_tutorial|data_tutorial|neural_networks_tutorial|nn_tutorial|optimizer_step_in_backward_tutorial|profiler_recipe|timer_quick_start|
  torch_compiler_set_stance_tutorial|gpu_direct_storage)\.py$' \
  sphinx-build -E -D plot_gallery=0 -b html . _build/html
```

**Comment by ribhuji** [comment]
@albanD the notebooks are fixed now, and the lintrunner will ensure this exact issue shouldn't re-occur.

Please run the workflows. Thanks!

**Comment by ribhuji** [comment]
Hey @albanD, bumping this up!

**Comment by ribhuji** [comment]
@albanD gentle reminder.

**Comment by ribhuji** [comment]
@albanD  gentle bump!
